### PR TITLE
Support for OFFSET in LogicalPlan

### LINKED ISSUE #2377: Add support for `OFFSET` in SQL query planner + logical plan
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
I would like to produce logical plans for queries that contain `OFFSET`.

**Describe the solution you'd like**
I would like to produce logical plans for queries that contain `OFFSET`.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by jdye64**
# Which issue does this PR close?

Closes #2377 

# What changes are included in this PR?
Modification to LogicalPlan and LogicalPlan builder to include the `Offset` variant that is already present in `sqlparser-rs`

# Are there any user-facing changes?
Yes, the proto definition has changed with the inclusion of the `Offset` variant

### CODE REVIEW

**Comment by jdye64** [comment]
I don't have permissions to add the label but this likely needs a `api change` label since the protobuf files are modified.

File: datafusion/core/src/sql/planner.rs [hunk 1]
```diff
@@ -293,6 +294,8 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
 
         let plan = self.order_by(plan, query.order_by)?;
 
+        let plan: LogicalPlan = self.offset(plan, query.offset)?;
```
Comment (andygrove) [inline datafusion/core/src/sql/planner.rs:297]: Could you add some tests demonstrating SQL queries with the various combinations of `OFFSET` and `LIMIT`

File: datafusion/core/src/sql/planner.rs [hunk 2]
```diff
@@ -1259,6 +1262,33 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
...
+        offset: Option<SQLOffset>,
+    ) -> Result<LogicalPlan> {
+        match offset {
+            Some(offset_expr) => {
+                let offset = match self.sql_to_rex(
+                    offset_expr.value,
+                    input.schema(),
+                    &mut HashMap::new(),
+                )? {
+                    Expr::Literal(ScalarValue::Int64(Some(offset))) => {
```
Comment (andygrove) [inline datafusion/core/src/sql/planner.rs:1278]: We should probably have a check that the offset is `>= 0` ?

**Review by andygrove — APPROVED** [review]
LGTM. Thanks @jdye64

**Comment by andygrove** [comment]
I filed two follow on issues:

- https://github.com/apache/arrow-datafusion/pull/2521
- https://github.com/apache/arrow-datafusion/pull/2522

**Review by yjshen — APPROVED** [review]

**Comment by ming535** [comment]
Hi, why `limit` be an input of `offset`? From my understanding, the `offset` is the starting point for where data is coming from, so I think `offset` should be an input of `limit` instead? @jdye64 @andygrove

**Comment by ming535** [comment]
Hi, why `limit` be an input of `offset`? From my understanding, the `offset` is the starting point for where data is coming from, so I think `offset` should be an input of `limit` instead? @jdye64 @andygrove

**Comment by alamb** [comment]
Hi @ming535  -- it depends on how the operators work

Of limit is the input to offset, I think you can get the right answers. 

For example, with `limit 10 offset 5` the idea would be that the `LIMIT`  produces 15 rows, and then the offset skips the first 5, resulting in 10 in total.

I realize this is not what the planner does now (it doesn't adjust the limit to 15, it uses 10). However, the plan also doesn't run as the physical plan for offset isn't completed. I will file a ticket for that

**Comment by alamb** [comment]
The physical part of the plan is tracked in https://github.com/apache/arrow-datafusion/issues/2551

**Pull request merged**